yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala

```diff
@@ -0,0 +1,158 @@
+package org.apache.spark.deploy.yarn.security
+
+import java.util.ServiceLoader
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.security.Credentials
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils
+
+/**
+ * A ConfigurableCredentialManager to manage all the registered credential providers and offer
+ * APIs for other modules to obtain credentials as well as renewal time. By default
+ * [[HDFSCredentialProvider]], [[HiveCredentialProvider]] and [[HBaseCredentialProvider]] will
+ * be loaded in, any plugged-in credential provider wants to be managed by
+ * ConfigurableCredentialManager needs to implement [[ServiceCredentialProvider]] interface and put
+ * into resources to be loaded by ServiceLoader.
+ *
+ * Also the specific credential provider is controlled by
+ * spark.yarn.security.credentials.{service}.enabled, it will not be loaded in if set to false.
+ */
+final class ConfigurableCredentialManager private[yarn] (sparkConf: SparkConf) extends Logging {
+  private val deprecatedProviderEnabledConfig = "spark.yarn.security.tokens.%s.enabled"
+  private val providerEnabledConfig = "spark.yarn.security.credentials.%s.enabled"
+
+  // Maintain all the registered credential providers
+  private val credentialProviders = mutable.HashMap[String, ServiceCredentialProvider]()
+
+  // Default crendetial providers that will be loaded automatically, unless specifically disabled.
+  private val defaultCredentialProviders = Map(
+    "hdfs" -> "org.apache.spark.deploy.yarn.security.HDFSCredentialProvider",
+    "hive" -> "org.apache.spark.deploy.yarn.security.HiveCredentialProvider",
+    "hbase" -> "org.apache.spark.deploy.yarn.security.HBaseCredentialProvider"
+  )
+
+  // AMDelegationTokenRenewer, this will only be create and started in the AM
+  private var _delegationTokenRenewer: AMDelegationTokenRenewer = null
+
+  // ExecutorDelegationTokenUpdater, this will only be created and started in the driver and
+  // executor side.
+  private var _delegationTokenUpdater: ExecutorDelegationTokenUpdater = null
+
+  def initialize(): Unit = {
+    val providers = ServiceLoader.load(classOf[ServiceCredentialProvider],
+      Utils.getContextOrSparkClassLoader).asScala
+
+    // Filter out credentials in which spark.yarn.security.credentials.{service}.enabled is false.
+    providers.filter { p =>
+      sparkConf.getOption(providerEnabledConfig.format(p.serviceName))
+        .orElse {
+          sparkConf.getOption(deprecatedProviderEnabledConfig.format(p.serviceName)).map { c =>
+            logWarning(s"${deprecatedProviderEnabledConfig.format(p.serviceName)} is deprecated, " +
+              s"using ${providerEnabledConfig.format(p.serviceName)} instead")
+            c
+          }
+        }
+        .getOrElse(defaultCredentialProviders.keySet.find(_ == p.serviceName).isDefined.toString)
```

Comment:
What's the purpose of this line? Seems like you want to enable internal providers and disable 3rd-party ones by default? Why?